src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,276 @@
+    }
+
+    struct FindByBegin : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & rhs) const
+        {
+            return value < rhs.range.begin;
+        }
+    };
+
+    struct FindByEndGreater : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & lhs) const
+        {
+            return value > lhs.range.end;
+        }
+    };
+
+    struct CompareByBegin : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.begin < rhs.range.begin;
+        }
+    };
+
+
+    struct CompareByEnd : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.end < rhs.range.end;
+        }
+    };
+};
+
+
+class PartRangesIntersectionsIndex
+{
+public:
+    void addPart(PartToRead part)
+    {
+        by_begin.insert(part);
+        by_end.insert(part);
+    }
+
+    size_t numberOfIntersectionsWith(PartBlockRange range)
+    {
+        /// Find the first one that starts with a larger coordinate
+        auto right_iter = std::upper_bound(by_begin.begin(), by_begin.end(), range.end, PartToRead::FindByBegin());
+
+        /// Find the last that ends with a lower coordinate
+        auto left_iter = std::upper_bound(by_end.rbegin(), by_end.rend(), range.begin, PartToRead::FindByEndGreater());
+
+        size_t result = by_begin.size();
+        result -= std::distance(left_iter, by_end.rend());
+        result -= std::distance(right_iter, by_begin.end());
+
+        return result;
+    }
+
+    bool checkPartIsSuitable(PartToRead part)
+    {
+        auto number = numberOfIntersectionsWith(part.range);
+        if (number == 0)
+            return true;
+        if (number == 1)
+        {
+            auto it = by_begin.find(part);
+            if (it != by_begin.end() && it->name == part.name)
+                return true;
+        }
+
+        return false;
+    }
+
+    String describe() const
+    {
+        String result = "[";
+        for (const auto & it : by_begin)
+            result += fmt::format("({} {} {}), ", it.name, it.range.begin, it.range.end);
+        result += "]";
+        return result;
+    }
+
+
+    void checkConsistencyOrThrow() const
+    {
+        Int64 prev_end = std::numeric_limits<Int64>::min();
+        for (const auto & it : by_begin)
+        {
+            if (it.range.begin < prev_end)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Intersecting marks");
+            prev_end = it.range.end;
+        }
+    }
+
+private:
+    std::multiset<PartToRead, PartToRead::CompareByBegin> by_begin;
+    std::multiset<PartToRead, PartToRead::CompareByEnd> by_end;
+};
+
+
+
+struct MarkRangesIntersectionsIndex
```

Comment:
I think we can create an independent container with half-intervals. It will represent a sorted set of non-intersected intervals, like `[1, 2), [4, 8), [10, 11)` (something like `[1, 2), [3, 4)` is not allowed and should be like `[1, 4)`)
I expected that we need only half-intervals intersection and negation operation.
Information for part may contain min and max possible range.
Index need to support `intervals to-do`, starting with `[min, max)`.
Every task should be intersected with `intervals to-do`. E.g.:
```
intervals to-do                      request               responce
[min, max)                            [4, 5), [8, 10)      [4, 5), [8, 10)
[min, 4), [5, 8), [10, max)    [2, 12)                 [2, 4), [5, 8) [12, max)
[min, 2), [12, max)
```
New `intervals to-do` are `intervals to-do & !request`. Negation is pretty easy, like `!([4, 5), [8, 10)) = [min, 4), [5, 8), [10, max)`.

Reply:
Done